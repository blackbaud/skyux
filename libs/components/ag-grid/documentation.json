{
  "$schema": "../manifest/documentation-schema.json",
  "groups": {
    "data-grid": {
      "docsIds": [
        "SkyAgGridModule",
        "SkyAgGridWrapperComponent",
        "SkyAgGridService",
        "SkyGetGridOptionsArgs",
        "SkyAgGridRowDeleteDirective",
        "SkyAgGridRowDeleteConfirmArgs",
        "SkyAgGridRowDeleteCancelArgs",
        "SkyAgGridCellEditorInitialAction",
        "SkyAgGridCellEditorUtils",
        "SkyCellType",
        "SkyAgGridHeaderInfo",
        "SkyAgGridHeaderParams",
        "SkyAgGridHeaderGroupInfo",
        "SkyAgGridHeaderGroupParams",
        "SkyAgGridAutocompleteProperties",
        "SkyAutocompleteProperties",
        "SkyAgGridCurrencyProperties",
        "SkyAgGridDatepickerProperties",
        "SkyDatepickerProperties",
        "SkyAgGridLookupProperties",
        "SkyAgGridNumberProperties",
        "SkyAgGridTextProperties",
        "SkyAgGridValidatorProperties",
        "AgGridDataGridBasicExampleComponent",
        "AgGridDataGridBasicMultiselectExampleComponent",
        "AgGridDataGridDataManagerExampleComponent",
        "AgGridDataGridDataManagerMultiselectExampleComponent",
        "AgGridDataGridInlineHelpExampleComponent",
        "AgGridDataGridPagingExampleComponent",
        "AgGridDataGridTemplateRefColumnExampleComponent",
        "AgGridDataGridTopScrollExampleComponent"
      ],
      "primaryDocsId": "SkyAgGridModule"
    },
    "data-entry-grid": {
      "docsIds": [
        "SkyAgGridModule",
        "SkyAgGridWrapperComponent",
        "SkyAgGridService",
        "SkyGetGridOptionsArgs",
        "SkyAgGridRowDeleteDirective",
        "SkyAgGridRowDeleteConfirmArgs",
        "SkyAgGridRowDeleteCancelArgs",
        "SkyAgGridCellEditorInitialAction",
        "SkyAgGridCellEditorUtils",
        "SkyCellType",
        "SkyAgGridHeaderInfo",
        "SkyAgGridHeaderParams",
        "SkyAgGridHeaderGroupInfo",
        "SkyAgGridHeaderGroupParams",
        "SkyAgGridAutocompleteProperties",
        "SkyAutocompleteProperties",
        "SkyAgGridCurrencyProperties",
        "SkyAgGridDatepickerProperties",
        "SkyDatepickerProperties",
        "SkyAgGridLookupProperties",
        "SkyAgGridNumberProperties",
        "SkyAgGridTextProperties",
        "SkyAgGridValidatorProperties",
        "AgGridDataEntryGridBasicExampleComponent",
        "AgGridDataEntryGridDataManagerAddedExampleComponent",
        "AgGridDataEntryGridFocusExampleComponent",
        "AgGridDataEntryGridInlineHelpExampleComponent"
      ],
      "primaryDocsId": "SkyAgGridModule"
    }
  }
}
