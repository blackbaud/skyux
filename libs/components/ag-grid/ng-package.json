{
  "$schema": "../../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../../dist/libs/components/ag-grid",
  "lib": {
    "entryFile": "src/index.ts",
    "styleIncludePaths": ["../../.."]
  },
  "inlineStyleLanguage": "scss",
  "allowedNonPeerDependencies": ["@skyux/icon"]
}
