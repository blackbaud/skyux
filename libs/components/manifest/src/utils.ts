import type { SkyManifestParentDefinition } from './types/base-def';
import type { SkyManifestDirectiveDefinition } from './types/directive-def';

/**
 * Whether the provided definition is a directive definition.
 * @internal
 */
export function isDirectiveDefinition(
  def: SkyManifestParentDefinition,
): def is SkyManifestDirectiveDefinition {
  return def.kind === 'component' || def.kind === 'directive';
}
