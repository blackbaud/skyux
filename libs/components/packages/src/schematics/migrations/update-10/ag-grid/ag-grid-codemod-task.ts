import {
  TaskConfiguration,
  TaskConfigurationGenerator,
} from '@angular-devkit/schematics';

// eslint-disable-next-line @cspell/spellchecker
export interface AgGridCodemodTaskOptions {
  /**
   * The source root of the project.
   */
  sourceRoot: string;
}

// eslint-disable-next-line @cspell/spellchecker
export class AgGridCodemodTask implements TaskConfigurationGenerator<unknown> {
  // eslint-disable-next-line @cspell/spellchecker
  toConfiguration(): TaskConfiguration<AgGridCodemodTaskOptions> {
    return {
      // eslint-disable-next-line @cspell/spellchecker
      name: 'run-ag-grid-codemod',
      options: {
        sourceRoot: 'sourceRoot',
      },
    };
  }
}
