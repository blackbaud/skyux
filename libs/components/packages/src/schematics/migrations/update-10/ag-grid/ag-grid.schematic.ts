import { Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
import { NodePackageInstallTask } from '@angular-devkit/schematics/tasks';
import {
  NodeDependencyType,
  addPackageJsonDependency,
  getPackageJsonDependency,
} from '@schematics/angular/utility/dependencies';

import { visitProjectFiles } from '../../../utility/visit-project-files';
import { getWorkspace } from '../../../utility/workspace';

const ANY_MODULE = '@ag-grid-community/';
const ENT_MODULE = '@ag-grid-enterprise/';
const AG_GRID = 'ag-grid-community';
const AG_GRID_ENT = 'ag-grid-enterprise';
const AG_GRID_NG = 'ag-grid-angular';
const AG_GRID_SKY = '@skyux/ag-grid';

const AG_GRID_VERSION = '^30.0.0';

/**
 * Check package.json for AG Grid dependencies.
 */
function checkAgGridDependency(tree: Tree, context: SchematicContext): boolean {
  const agGridDependency = getPackageJsonDependency(tree, AG_GRID);
  const agGridDependencyEnt = getPackageJsonDependency(tree, AG_GRID_ENT);
  const agGridDependencyNg = getPackageJsonDependency(tree, AG_GRID_NG);
  if (agGridDependency || agGridDependencyEnt || agGridDependencyNg) {
    if (!agGridDependency) {
      // Missing peer dependency.
      addPackageJsonDependency(tree, {
        name: AG_GRID,
        type: NodeDependencyType.Default,
        version: AG_GRID_VERSION,
      });
      context.addTask(new NodePackageInstallTask());
    }
    return true;
  }

  const packageJson = tree.readJson('package.json') as Record<
    string,
    Record<string, string>
  >;

  return ['dependencies', 'devDependencies'].some((dep) =>
    Object.keys((packageJson && packageJson[dep]) || {}).some(
      (dep) => dep.startsWith(ANY_MODULE) || dep.startsWith(ENT_MODULE),
    ),
  );
}

/**
 * Column API renamed getSecondaryColumns / setSecondaryColumns to getSecondaryPivotColumns / setSecondaryPivotColumns.
 */
function renameColumnApiFunctionsInCode(updatedContent: string): string {
  if (updatedContent.includes('etSecondaryColumns(')) {
    updatedContent = updatedContent.replace(
      /(?<=columnApi\s*)\.(get|set)SecondaryColumns\(/g,
      (_, x) => `.${x}PivotResultColumns(`,
    );
  }
  return updatedContent;
}

/**
 * Update renamed grid options in code.
 */
function renameGridOptionsInCode(updatedContent: string): string {
  if (
    updatedContent.match(/gridOptions/i) &&
    (updatedContent.includes('suppressCellSelection') ||
      updatedContent.includes('getRowNodeId') ||
      updatedContent.includes('enterMovesDown'))
  ) {
    updatedContent = updatedContent
      .replace(/\bsuppressCellSelection\b/g, 'suppressCellFocus')
      .replace(/\bgetRowNodeId\b/g, 'getRowId')
      .replace(/\benterMovesDown(?=\b|AfterEdit)/g, 'enterNavigatesVertically');
  }
  return updatedContent;
}

/**
 * Switch charPress to eventKey.
 */
function renameCharPress(updatedContent: string): string {
  if (updatedContent.match(/\bcharPress\b(?!: undefined)/)) {
    updatedContent = updatedContent.replace(
      /\bcharPress\b(?!: undefined)/g,
      'eventKey',
    );
  }
  return updatedContent;
}

/**
 * Switch cellRendererFramework to cellRenderer.
 */
function renameCellRendererFramework(updatedContent: string): string {
  if (updatedContent.includes('cellRendererFramework')) {
    updatedContent = updatedContent.replace(
      /\bcellRendererFramework\b/g,
      'cellRenderer',
    );
  }
  return updatedContent;
}

/**
 * Check if the file includes any AG Grid imports.
 */
function includesAgGrid(updatedContent: string): boolean {
  return (
    updatedContent.includes(AG_GRID) ||
    updatedContent.includes(AG_GRID_ENT) ||
    updatedContent.includes(AG_GRID_SKY) ||
    updatedContent.includes(ANY_MODULE) ||
    updatedContent.includes(ENT_MODULE)
  );
}

/**
 * Visit all files and apply the changes.
 */
async function updateSourceFiles(
  tree: Tree,
  context: SchematicContext,
): Promise<void> {
  const warned: string[] = [];

  function warnOnce(message: string) {
    if (!warned.includes(message)) {
      warned.push(message);
      context.logger.warn(message);
    }
  }

  const { workspace } = await getWorkspace(tree);
  workspace.projects.forEach((project) => {
    visitProjectFiles(tree, project.sourceRoot || project.root, (filePath) => {
      // If the file is not a TypeScript file, we can skip it.
      if (!filePath.endsWith('.ts')) {
        return;
      }
      const content = tree.readText(filePath);
      if (!content || !includesAgGrid(content)) {
        return;
      }
      let updatedContent = content;

      // Prompt the user to moderate the use of AG Grid modules
      if (
        updatedContent.includes(ANY_MODULE) ||
        updatedContent.includes(ENT_MODULE)
      ) {
        warnOnce(
          `\n
          AG Grid recommends not mixing module and package imports.
          https://ag-grid.com/angular-data-grid/modules/\n\n`,
        );
      }

      updatedContent = renameCellRendererFramework(updatedContent);
      updatedContent = renameCharPress(updatedContent);
      updatedContent = renameColumnApiFunctionsInCode(updatedContent);
      updatedContent = renameGridOptionsInCode(updatedContent);

      if (updatedContent !== content) {
        tree.overwrite(filePath, updatedContent);
      }
    });
  });
}

/**
 * Upgrade to AG Grid 30 and address three breaking changes:
 *
 * - Add missing peer dependency
 * - Warn about mixing modules and packages
 * - Column API renamed getSecondaryColumns / setSecondaryColumns
 * - Grid option renamed suppressCellSelection and getRowNodeId
 * - Rename charPress to eventKey
 * - Rename cellRendererFramework to cellRenderer
 *
 * Also advise against mixing modules and packages.
 */
export default function (): Rule {
  return async (tree: Tree, context: SchematicContext): Promise<void> => {
    const hasAgGrid = checkAgGridDependency(tree, context);

    // AG Grid is not installed, so we don't need to do anything.
    if (!hasAgGrid) {
      return;
    }

    await updateSourceFiles(tree, context);
  };
}
