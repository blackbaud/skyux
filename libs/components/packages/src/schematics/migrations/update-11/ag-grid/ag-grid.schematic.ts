import { Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
import {
  NodePackageInstallTask,
  RunSchematicTask,
} from '@angular-devkit/schematics/tasks';
import { parseSourceFile } from '@angular/cdk/schematics';
import ts from '@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript';
import {
  NodeDependencyType,
  addPackageJsonDependency,
  getPackageJsonDependency,
} from '@schematics/angular/utility/dependencies';

import { swapImportedClass } from '../../../utility/typescript/swap-imported-class';
import { visitProjectFiles } from '../../../utility/visit-project-files';
import { getWorkspace } from '../../../utility/workspace';

import { eventTypeStrings } from './event-type-strings/event-type-strings';
import { switchToSetGridOption } from './switch-to-set-grid-option/switch-to-set-grid-option';

const ANY_MODULE = '@ag-grid-community/';
const ENT_MODULE = '@ag-grid-enterprise/';
const AG_GRID = 'ag-grid-community';
const AG_GRID_ENT = 'ag-grid-enterprise';
const AG_GRID_NG = 'ag-grid-angular';
const AG_GRID_SKY = '@skyux/ag-grid';

const AG_GRID_VERSION = '^32.2.2';

/**
 * Check package.json for AG Grid dependencies.
 */
function checkAgGridDependency(tree: Tree, context: SchematicContext): boolean {
  const agGridDependency = getPackageJsonDependency(tree, AG_GRID);
  const agGridDependencyEnt = getPackageJsonDependency(tree, AG_GRID_ENT);
  const agGridDependencyNg = getPackageJsonDependency(tree, AG_GRID_NG);
  if (agGridDependency || agGridDependencyEnt || agGridDependencyNg) {
    if (!agGridDependency) {
      // Missing peer dependency.
      addPackageJsonDependency(tree, {
        name: AG_GRID,
        type: NodeDependencyType.Default,
        version: AG_GRID_VERSION,
      });
      context.addTask(new NodePackageInstallTask());
    }
    return true;
  }

  const packageJson = tree.readJson('package.json') as Record<
    string,
    Record<string, string>
  >;

  return ['dependencies', 'devDependencies'].some((dep) =>
    Object.keys((packageJson && packageJson[dep]) || {}).some(
      (dep) => dep.startsWith(ANY_MODULE) || dep.startsWith(ENT_MODULE),
    ),
  );
}

/**
 * If available, switch gridOptions.api and gridOptions.columnApi to gridApi.
 */
function swapGridOptionsApiToGridApi(tree: Tree, path: string): void {
  const content = tree.readText(path);
  const recorder = tree.beginUpdate(path);
  if (
    content.includes('this.gridApi') &&
    (content.includes('this.gridOptions.api') ||
      content.includes('this.gridOptions.columnApi'))
  ) {
    const instances = content.matchAll(
      /\bthis\.gridOptions\.(api|columnApi)\b/g,
    );
    for (const instance of instances) {
      recorder.remove(instance.index, instance[0].length);
      recorder.insertRight(instance.index, 'this.gridApi');
    }
  }
  if (content.includes('columnApi = ')) {
    const instances = content.matchAll(
      /(?<=\bthis\.[_#]?[A-Za-z]*[Cc]olumnApi = \w+\.|event\.)columnApi(?=[.;])/g,
    );
    for (const instance of instances) {
      recorder.remove(instance.index, instance[0].length);
      recorder.insertRight(instance.index, 'api');
    }
  }
  tree.commitUpdate(recorder);
  if (content.includes('ColumnApi')) {
    swapImportedClass(tree, path, parseSourceFile(tree, path), [
      {
        classNames: {
          ColumnApi: 'GridApi',
        },
        moduleName: 'ag-grid-community',
      },
    ]);
  }
}

/**
 * Check if the file includes any AG Grid imports.
 */
function includesAgGrid(updatedContent: string): boolean {
  return (
    updatedContent.includes(AG_GRID) ||
    updatedContent.includes(AG_GRID_ENT) ||
    updatedContent.includes(AG_GRID_SKY)
  );
}

function swapClasses(tree: Tree, filePath: string) {
  const sourceFile = parseSourceFile(tree, filePath);
  swapImportedClass(tree, filePath, sourceFile, [
    {
      classNames: {
        Column: 'AgColumn',
        Beans: 'BeanCollection',
      },
      moduleName: 'ag-grid-community',
      filter: (node: ts.Identifier) => {
        const parent = node.parent;
        return (
          node.text !== 'Column' ||
          (parent && ts.isNewExpression(parent) && parent.expression === node)
        );
      },
    },
  ]);
}

/**
 * Visit all files and apply the changes.
 */
async function updateSourceFiles(
  tree: Tree,
  context: SchematicContext,
): Promise<void> {
  const warned: string[] = [];

  function warnOnce(message: string) {
    if (!warned.includes(message)) {
      warned.push(message);
      context.logger.warn(message);
    }
  }

  const { workspace } = await getWorkspace(tree);
  workspace.projects.forEach((project) => {
    context.logger.debug(
      `Running SKY UX AG Grid updates within ${project.sourceRoot || project.root}.`,
    );
    visitProjectFiles(tree, project.sourceRoot || project.root, (filePath) => {
      // If the file is not a TypeScript file, we can skip it.
      if (!filePath.endsWith('.ts') || filePath.includes('schematics')) {
        return;
      }
      const content = tree.readText(filePath);
      if (!content || !includesAgGrid(content)) {
        return;
      }

      // Prompt the user to moderate the use of AG Grid modules
      if (content.includes(ANY_MODULE) || content.includes(ENT_MODULE)) {
        warnOnce(
          `\n
          AG Grid recommends not mixing module and package imports.
          https://ag-grid.com/angular-data-grid/modules/\n\n`,
        );
      }

      eventTypeStrings(tree, filePath);
      swapClasses(tree, filePath);
      swapGridOptionsApiToGridApi(tree, filePath);
      switchToSetGridOption(tree, filePath);
    });
  });
}

/**
 * Upgrade to AG Grid 32 and address three breaking changes:
 *
 * - Switch Event Types to Strings
 * - Switch gridOptions.api and gridOptions.columnApi to gridApi
 * - Switch setQuickFilter to updateGridOptions
 * - Switch Beans references to use BeanCollection
 * - Switch Column references to use AgColumn
 *
 * Also advise against mixing modules and packages.
 */
export default function (): Rule {
  return async (tree: Tree, context: SchematicContext): Promise<void> => {
    const hasAgGrid = checkAgGridDependency(tree, context);

    // AG Grid is not installed, so we don't need to do anything.
    if (!hasAgGrid) {
      context.logger.debug(`AG Grid is not installed.`);
      return;
    }

    await updateSourceFiles(tree, context);
    const { workspace } = await getWorkspace(tree);
    for (const project of workspace.projects.values()) {
      const sourceRoot = project.sourceRoot || `${project.root}/src`;
      context.logger.debug(
        `Scheduling AG Grid code modifications for ${sourceRoot}.`,
      );
      context.addTask(
        new RunSchematicTask('@skyux/packages', 'ag-grid-migrate', {
          sourceRoot,
        }),
      );
    }
  };
}
