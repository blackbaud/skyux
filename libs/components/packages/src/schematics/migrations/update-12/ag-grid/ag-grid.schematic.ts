import { Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
import {
  NodePackageInstallTask,
  RunSchematicTask,
} from '@angular-devkit/schematics/tasks';
import {
  NodeDependencyType,
  addPackageJsonDependency,
  getPackageJsonDependency,
} from '@schematics/angular/utility/dependencies';

import { visitProjectFiles } from '../../../utility/visit-project-files';
import { getWorkspace } from '../../../utility/workspace';

const ANY_MODULE = '@ag-grid-community/';
const ENT_MODULE = '@ag-grid-enterprise/';
const AG_GRID = 'ag-grid-community';
const AG_GRID_ENT = 'ag-grid-enterprise';
const AG_GRID_NG = 'ag-grid-angular';
const AG_GRID_SKY = '@skyux/ag-grid';

const AG_GRID_VERSION = '^32.3.3';

/**
 * Check package.json for AG Grid dependencies.
 */
function checkAgGridDependency(tree: Tree, context: SchematicContext): boolean {
  const agGridDependency = getPackageJsonDependency(tree, AG_GRID);
  const agGridDependencyEnt = getPackageJsonDependency(tree, AG_GRID_ENT);
  const agGridDependencyNg = getPackageJsonDependency(tree, AG_GRID_NG);
  if (agGridDependency || agGridDependencyEnt || agGridDependencyNg) {
    if (!agGridDependency) {
      // Missing peer dependency.
      addPackageJsonDependency(tree, {
        name: AG_GRID,
        type: NodeDependencyType.Default,
        version: AG_GRID_VERSION,
      });
      context.addTask(new NodePackageInstallTask());
    }
    return true;
  }

  const packageJson = tree.readJson('package.json') as Record<
    string,
    Record<string, string>
  >;

  return ['dependencies', 'devDependencies'].some((dep) =>
    Object.keys((packageJson && packageJson[dep]) || {}).some(
      (dep) => dep.startsWith(ANY_MODULE) || dep.startsWith(ENT_MODULE),
    ),
  );
}

/**
 * Check if the file includes any AG Grid imports.
 */
function includesAgGrid(updatedContent: string): boolean {
  return (
    updatedContent.includes(AG_GRID) ||
    updatedContent.includes(AG_GRID_ENT) ||
    updatedContent.includes(AG_GRID_SKY)
  );
}

/**
 * Visit all files and apply the changes.
 */
async function updateSourceFiles(
  tree: Tree,
  context: SchematicContext,
): Promise<void> {
  const warned: string[] = [];

  function warnOnce(message: string) {
    if (!warned.includes(message)) {
      warned.push(message);
      context.logger.warn(message);
    }
  }

  const { workspace } = await getWorkspace(tree);
  workspace.projects.forEach((project) => {
    context.logger.debug(
      `Running SKY UX AG Grid updates within ${project.sourceRoot || project.root}.`,
    );
    visitProjectFiles(tree, project.sourceRoot || project.root, (filePath) => {
      // If the file is not a TypeScript file, we can skip it.
      if (!filePath.endsWith('.ts') || filePath.includes('schematics')) {
        return;
      }
      const content = tree.readText(filePath);
      if (!content || !includesAgGrid(content)) {
        return;
      }

      // Prompt the user to moderate the use of AG Grid modules
      if (content.includes(ANY_MODULE) || content.includes(ENT_MODULE)) {
        warnOnce(
          `\n
          AG Grid recommends not mixing module and package imports.
          https://ag-grid.com/angular-data-grid/modules/\n\n`,
        );
      }
    });
  });
}

/**
 * Upgrade to AG Grid 32 and address breaking changes:
 *
 * Also advise against mixing modules and packages.
 */
export default function (): Rule {
  return async (tree: Tree, context: SchematicContext): Promise<void> => {
    const hasAgGrid = checkAgGridDependency(tree, context);

    // AG Grid is not installed, so we don't need to do anything.
    if (!hasAgGrid) {
      context.logger.debug(`AG Grid is not installed.`);
      return;
    }

    await updateSourceFiles(tree, context);
    const { workspace } = await getWorkspace(tree);
    for (const project of workspace.projects.values()) {
      const sourceRoot = project.sourceRoot || `${project.root}/src`;
      context.logger.debug(
        `Scheduling AG Grid code modifications for ${sourceRoot}.`,
      );
      context.addTask(
        new RunSchematicTask('@skyux/packages', 'ag-grid-migrate', {
          sourceRoot,
        }),
      );
    }
  };
}
