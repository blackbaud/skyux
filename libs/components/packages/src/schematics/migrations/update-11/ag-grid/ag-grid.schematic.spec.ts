import { Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';

import fs from 'fs-extra';
import { joinPathFragments } from 'nx/src/utils/path';
import { workspaceRoot } from 'nx/src/utils/workspace-root';

const UPDATE_TO_VERSION = '32.2.2';

describe('ag-grid.schematic', () => {
  const runner = new SchematicTestRunner(
    'schematics',
    require.resolve('../../migration-collection.json'),
  );
  const angularJson = {
    version: 1,
    projects: {
      test: {
        projectType: 'application',
        root: '',
        architect: {},
      },
    },
  };

  function setupTest(
    packageJson: Record<string, Record<string, string>> = {},
  ): { tree: Tree } {
    const tree = Tree.empty();
    tree.create('/angular.json', JSON.stringify(angularJson));
    tree.create('/package.json', JSON.stringify(packageJson));
    return { tree };
  }

  it('should test the current version', () => {
    const packageJson = fs.readJSONSync(
      joinPathFragments(workspaceRoot, 'package.json'),
    );
    expect(packageJson.dependencies['ag-grid-community']).toBe(
      UPDATE_TO_VERSION,
    );
  });

  it('should work', async () => {
    expect.assertions(1);
    const { tree } = setupTest({
      dependencies: {
        'ag-grid-community': UPDATE_TO_VERSION,
        'ag-grid-angular': UPDATE_TO_VERSION,
        'ag-grid-enterprise': UPDATE_TO_VERSION,
      },
    });
    await runner.runSchematic('ag-grid', {}, tree);
    expect(JSON.parse(tree.readText('/package.json'))).toEqual({
      dependencies: {
        'ag-grid-community': UPDATE_TO_VERSION,
        'ag-grid-angular': UPDATE_TO_VERSION,
        'ag-grid-enterprise': UPDATE_TO_VERSION,
      },
    });
  });

  it('should add missing peer', async () => {
    expect.assertions(1);
    const { tree } = setupTest({
      dependencies: {
        'ag-grid-angular': UPDATE_TO_VERSION,
      },
    });
    await runner.runSchematic('ag-grid', {}, tree);
    expect(JSON.parse(tree.readText('/package.json'))).toEqual({
      dependencies: {
        'ag-grid-community': `^${UPDATE_TO_VERSION}`,
        'ag-grid-angular': UPDATE_TO_VERSION,
      },
    });
  });

  it('should noop', async () => {
    expect.assertions(1);
    const { tree } = setupTest({
      dependencies: {
        other: '27.1.1',
      },
    });
    await runner.runSchematic('ag-grid', {}, tree);
    expect(JSON.parse(tree.readText('/package.json'))).toEqual({
      dependencies: {
        other: '27.1.1',
      },
    });
  });

  it('should use this.gridApi if possible', async () => {
    expect.assertions(1);
    const { tree } = setupTest({
      dependencies: {
        '@skyux/ag-grid': '0.0.0',
        'ag-grid-community': UPDATE_TO_VERSION,
        'ag-grid-angular': UPDATE_TO_VERSION,
      },
    });
    tree.create(
      'src/app/grid.component.ts',
      `
        import { Component } from '@angular/core';
        import { GridApi, GridOptions } from 'ag-grid-community';

        @Component()
        export class GridComponent {
          public onClick(): void {
            this.gridApi.deselectAll();
            this.gridOptions.api.deselectAll();
          }
        }`,
    );
    tree.create(
      'src/app/ent.component.ts',
      `
        import { Component } from '@angular/core';
        import { GridApi, GridOptions } from 'ag-grid-enterprise';

        @Component()
        export class GridComponent {
          public onClick(): void {
            this.gridApi.deselectAll();
            this.gridOptions.api.deselectAll();
          }
        }`,
    );
    tree.create(
      'src/app/sky-ag-grid.component.ts',
      `
        import { Component } from '@angular/core';
        import { SkyAgGridService } from '@skyux/ag-grid';`,
    );
    tree.create(
      'src/app/unrelated.component.ts',
      `import { Component } from '@angular/core';`,
    );
    tree.create(
      'src/app/grid-with-column-api.component.ts',
      `
        import { Component } from '@angular/core';
        import { GridApi, GridOptions } from 'ag-grid-community';

        @Component()
        export class GridComponent {
          public onClick(): void {
            this.gridApi.deselectAll();
            this.gridOptions.columnApi.deselectAll();
          }
        }`,
    );
    await runner.runSchematic('ag-grid', {}, tree);
    expect(tree.readText('src/app/grid.component.ts')).toMatchSnapshot();
  });

  it('should update this.columnApi', async () => {
    expect.assertions(1);
    const { tree } = setupTest({
      dependencies: {
        '@skyux/ag-grid': '0.0.0',
        'ag-grid-community': UPDATE_TO_VERSION,
        'ag-grid-angular': UPDATE_TO_VERSION,
      },
    });
    tree.create(
      'src/app/grid.component.ts',
      `
        import { Component } from '@angular/core';
        import { ColumnApi, GridApi, GridOptions, GridReadyEvent } from 'ag-grid-community';

        @Component()
        export class GridComponent {
          #gridReady(event: GridReadyEvent): void {
            this.columnApi = event.columnApi;
            this.#columnApi = event.columnApi;
            this._columnApi = event.columnApi;
            event.columnApi.getColumnState();
            event.columnApi.applyColumnState({ state: [] });
          }
        }`,
    );
    await runner.runSchematic('ag-grid', {}, tree);
    expect(tree.readText('src/app/grid.component.ts')).toMatchSnapshot();
  });

  it('should warn about mixing modules and packages', async () => {
    expect.assertions(1);
    const { tree } = setupTest({
      dependencies: {
        '@ag-grid-community/core': UPDATE_TO_VERSION,
        '@skyux/ag-grid': '0.0.0',
        'ag-grid-community': UPDATE_TO_VERSION,
        'ag-grid-angular': UPDATE_TO_VERSION,
      },
    });
    tree.create(
      'src/app/app.component.ts',
      `
        import { SkyAgGridService } from '@skyux/ag-grid';
        import { GridOptions } from '@ag-grid-community/core';

        export class AppComponent {
          public options: GridOptions;
          #agGridService: SkyAgGridService;

          constructor(agGridService: SkyAgGridService) {
            this.#agGridService = agGridService;
            let customOptions: Partial<GridOptions> = {};
            customOptions.suppressCellSelection = true;
            this.options = this.agGridService.getGridOptions({
              ...customOptions,
              suppressCellSelection: true
            });
          }
        }`,
    );
    tree.create(
      'src/app/ent.component.ts',
      `
        import { SkyAgGridService } from '@skyux/ag-grid';
        import { GridOptions } from '@ag-grid-enterprise/core';

        export class AppComponent {
          public options: GridOptions;
          #agGridService: SkyAgGridService;

          constructor(agGridService: SkyAgGridService) {
            this.#agGridService = agGridService;
            let customOptions: Partial<GridOptions> = {};
            customOptions.suppressCellSelection = true;
            this.options = this.agGridService.getGridOptions({
              ...customOptions,
              suppressCellSelection: true
            });
          }
        }`,
    );
    tree.create(
      'src/app/other.component.ts',
      `import { ColDef } from '@ag-grid-community/core';`,
    );
    await runner.runSchematic('ag-grid', {}, tree);
    expect(tree.readText('src/app/other.component.ts')).toEqual(
      `import { ColDef } from '@ag-grid-community/core';`,
    );
  });

  it('should switch `new Column` to `new AgColumn`', async () => {
    const { tree } = setupTest({
      dependencies: {
        '@skyux/ag-grid': '0.0.0',
        'ag-grid-community': UPDATE_TO_VERSION,
        'ag-grid-angular': UPDATE_TO_VERSION,
      },
    });
    tree.create(
      'src/app/app.component.ts',
      `
        import { SkyAgGridService } from '@skyux/ag-grid';
        import { Column } from 'ag-grid-community';

        export class AppComponent {
          public column: Column;
        }`,
    );
    tree.create(
      'src/app/app.component.spec.ts',
      `
        import { Column } from 'ag-grid-community';

        if('should work', () => {
          const column = new Column();
        });`,
    );

    await runner.runSchematic('ag-grid', {}, tree);

    expect(tree.readText('src/app/app.component.ts')).toEqual(
      `
        import { SkyAgGridService } from '@skyux/ag-grid';
        import { Column } from 'ag-grid-community';

        export class AppComponent {
          public column: Column;
        }`,
    );
    expect(tree.readText('src/app/app.component.spec.ts')).toEqual(
      `
        import {  AgColumn } from 'ag-grid-community';

        if('should work', () => {
          const column = new AgColumn();
        });`,
    );
  });

  it('should switch `Beans` to `BeanCollection`', async () => {
    const { tree } = setupTest({
      dependencies: {
        '@skyux/ag-grid': '0.0.0',
        'ag-grid-community': UPDATE_TO_VERSION,
        'ag-grid-angular': UPDATE_TO_VERSION,
      },
    });
    tree.create(
      'src/app/app.component.ts',
      `
        import { SkyAgGridService } from '@skyux/ag-grid';
        import { Beans } from 'ag-grid-community';

        export class AppComponent {
          public beans: Beans;
        }`,
    );
    tree.create(
      'src/app/app.component.spec.ts',
      `
        import { Beans } from 'ag-grid-community';

        if('should work', () => {
          const beans = new Beans();
        });`,
    );

    await runner.runSchematic('ag-grid', {}, tree);

    expect(tree.readText('src/app/app.component.ts')).toEqual(
      `
        import { SkyAgGridService } from '@skyux/ag-grid';
        import {  BeanCollection } from 'ag-grid-community';

        export class AppComponent {
          public beans: BeanCollection;
        }`,
    );
    expect(tree.readText('src/app/app.component.spec.ts')).toEqual(
      `
        import {  BeanCollection } from 'ag-grid-community';

        if('should work', () => {
          const beans = new BeanCollection();
        });`,
    );
  });
});
