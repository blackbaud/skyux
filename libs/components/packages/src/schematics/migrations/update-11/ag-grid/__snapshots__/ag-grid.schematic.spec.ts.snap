// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ag-grid.schematic should update this.columnApi 1`] = `
"
        import { Component } from '@angular/core';
        import {  GridApi, GridOptions, GridReadyEvent } from 'ag-grid-community';

        @Component()
        export class GridComponent {
          #gridReady(event: GridReadyEvent): void {
            this.columnApi = event.api;
            this.#columnApi = event.api;
            this._columnApi = event.api;
            event.api.getColumnState();
            event.api.applyColumnState({ state: [] });
          }
        }"
`;

exports[`ag-grid.schematic should use this.gridApi if possible 1`] = `
"
        import { Component } from '@angular/core';
        import { GridApi, GridOptions } from 'ag-grid-community';

        @Component()
        export class GridComponent {
          public onClick(): void {
            this.gridApi.deselectAll();
            this.gridApi.deselectAll();
          }
        }"
`;
